### PR TITLE
Added cluster states 

### PR DESCRIPTION
**Pull request by mdogan**
Introducing cluster states `{ACTIVE, FROZEN, SHUTTING_DOWN}` and `Cluster.changeClusterState()` API to transactionally change state in the whole cluster. 

- `ACTIVE` is normal operating state of the cluster. All operations will be allowed to run, nothing will be restricted in the cluster.
- `FROZEN` is to support rolling upgrade. In `FROZEN` state, cluster members and partition assignments will be frozen. No new member will be allowed to join, only the ones left in this state will be allowed.
- `SHUTTING_DOWN` is shutdown preparation state. This state add more restrictions to the `FROZEN` state. Additional to the `FROZEN` state, no operations will be allowed to run in the cluster, except the ones marked as `AllowedDuringShutdown`. Also, cluster state cannot be changed back to `FROZEN` or `ACTIVE` once state becomes `SHUTTING_DOWN`.

Split-brain merge process is also adapted to use `FROZEN` cluster-state before starting merge.

_This is built on https://github.com/hazelcast/hazelcast/pull/6237 and part of required changes for hot-restart._

_See more in commit messages and  [Hot Restart Cluster Shutdown Doc](https://hazelcast.atlassian.net/wiki/pages/viewpage.action?spaceKey=EN&title=Hot+Restart+Persistence+Design#HotRestartPersistenceDesign-ShuttingDownWholeCluster)_

### CODE REVIEW

File: hazelcast-ra/hazelcast-jca/src/test/java/com/hazelcast/jca/HazelcastConnectionImplTest.java [hunk 1]
```diff
@@ -191,7 +190,7 @@ public void getJobTracker() {
     @Test
     public void getCluster() {
         Cluster cluster = connection.getCluster();
-        assertNotSame(cluster, hz.getCluster());
```
Comment (pveentjer) [inline hazelcast-ra/hazelcast-jca/src/test/java/com/hazelcast/jca/HazelcastConnectionImplTest.java]: Just a question: what was returned in the old situation? Some kind of protective wrapper?

File: hazelcast/src/main/java/com/hazelcast/cluster/ClusterState.java [hunk 2]
```diff
@@ -0,0 +1,132 @@
+/*
```
Comment (pveentjer) [inline hazelcast/src/main/java/com/hazelcast/cluster/ClusterState.java]: Extensive documentation! :)

File: hazelcast/src/main/java/com/hazelcast/cluster/ClusterState.java [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+     * <li>
+     * It's not allowed to go back any other state, once {@code ClusterState} is changed to
+     * {@code SHUTTING_DOWN}. From now on, only possible/allowed action is shutting down the instance.
+     * </li>
+     * </ul>
+     */
+    SHUTTING_DOWN,
+
+    /**
+     * Shows that ClusterState in transition.
```
Comment (pveentjer) [inline hazelcast/src/main/java/com/hazelcast/cluster/ClusterState.java]: Can you add a few lines of comments when the transition state happens?

File: hazelcast/src/main/java/com/hazelcast/cluster/impl/AbstractJoiner.java [hunk 4]
```diff
@@ -176,7 +168,7 @@ boolean shouldMerge(JoinMessage joinMessage) {
 
         ClusterServiceImpl clusterService = node.getClusterService();
         try {
-            boolean validJoinRequest = clusterService.validateJoinMessage(joinMessage);
+            boolean validJoinRequest = clusterService.getClusterJoinManager().validateJoinMessage(joinMessage);
```
Comment (pveentjer) [inline hazelcast/src/main/java/com/hazelcast/cluster/impl/AbstractJoiner.java]: Can't you get the clusterJoinManager dependency in the constructor?

File: hazelcast/src/main/java/com/hazelcast/cluster/impl/AbstractJoiner.java [hunk 5]
```diff
@@ -293,40 +285,47 @@ public void reset() {
...
-            }
-        }
-
-        waitWithDeadline(calls, 3, TimeUnit.SECONDS, whileWaitMergeExceptionHandler);
-
-        final PrepareMergeOperation prepareMergeOperation = new PrepareMergeOperation(targetAddress);
-        prepareMergeOperation.setNodeEngine(node.nodeEngine).setService(node.getClusterService())
-                .setOperationResponseHandler(createEmptyResponseHandler());
-        operationService.runOperationOnCallingThread(prepareMergeOperation);
+        ClusterServiceImpl clusterService = node.clusterService;
```
Comment (pveentjer) [inline hazelcast/src/main/java/com/hazelcast/cluster/impl/AbstractJoiner.java]: Can't you get this dependency in the constructor?

File: hazelcast/src/main/java/com/hazelcast/cluster/impl/AbstractJoiner.java [hunk 6]
```diff
@@ -293,40 +285,47 @@ public void reset() {
...
+                return true;
+            } catch (Exception e) {
+                String error = e.getClass().getName() + ": " + e.getMessage();
+                logger.warning("While freezing cluster state! " + error);
+            }
+
+            try {
+                Thread.sleep(1000);
+            } catch (InterruptedException e) {
+                logger.warning("Interrupted while preparing cluster for merge!");
```
Comment (pveentjer) [inline hazelcast/src/main/java/com/hazelcast/cluster/impl/AbstractJoiner.java]: Probably best to restore the interrupt flag; now it gets lost.

File: hazelcast/src/main/java/com/hazelcast/cluster/impl/ClusterProxy.java [hunk 7]
```diff
@@ -1,56 +0,0 @@
-/*
```
Comment (pveentjer) [inline hazelcast/src/main/java/com/hazelcast/cluster/impl/ClusterProxy.java]: Ok.. this is the wrapper thing that is deleted.

File: hazelcast-ra/hazelcast-jca/src/test/java/com/hazelcast/jca/HazelcastConnectionImplTest.java (hunk 1 above)
Comment (mdogan) [inline hazelcast-ra/hazelcast-jca/src/test/java/com/hazelcast/jca/HazelcastConnectionImplTest.java]: There was a `ClusterProxy` instance wrapping cluster service itself. I just removed it and made cluster service an implementation of Cluster interface.

File: hazelcast/src/main/java/com/hazelcast/cluster/ClusterState.java (hunk 3 above)
Comment (mdogan) [inline hazelcast/src/main/java/com/hazelcast/cluster/ClusterState.java]: done

File: hazelcast/src/main/java/com/hazelcast/cluster/impl/AbstractJoiner.java (hunk 4 above)
Comment (mdogan) [inline hazelcast/src/main/java/com/hazelcast/cluster/impl/AbstractJoiner.java]: done

File: hazelcast/src/main/java/com/hazelcast/cluster/impl/AbstractJoiner.java (hunk 5 above)
Comment (mdogan) [inline hazelcast/src/main/java/com/hazelcast/cluster/impl/AbstractJoiner.java]: done

File: hazelcast/src/main/java/com/hazelcast/cluster/impl/AbstractJoiner.java (hunk 6 above)
Comment (mdogan) [inline hazelcast/src/main/java/com/hazelcast/cluster/impl/AbstractJoiner.java]: done

**Comment by mtopolnik** [comment]
:+1:

**Comment by metanet** [comment]
:+1:

**Comment by gurbuzali** [comment]
:+1:

**Pull request merged**